### PR TITLE
S3 Block Public Access Fix

### PR DESCRIPTION
**Pull request by tgelpi**
#### Description
Starting in April 2023, Amazon S3 will introduce two new default bucket security settings by automatically enabling S3 Block Public Access and disabling S3 access control lists (ACLs) for all new S3 buckets. Once complete, these defaults will apply to all new buckets regardless of how they are created, including AWS CLI, APIs, SDKs, and AWS CloudFormation.

This AWS change results in the error **AccessControlListNotSupported: The bucket does not allow ACLs** when  creating a new EKS environment.

#### Special notes for the reviewer(s)
This change replaces the use of **_aws_s3_bucket_acl_** resource with the **_aws_s3_bucket_ownership_** resource. The rule for the ownership is **_BucketOwnerEnforced_**.

In the event ACL is still the preferred bucket permission type, you can set a new variable `enable_acl=true.` 

The tests were limited and run using both the **_Vault_** and **_AWS Secret Manager_** cluster config. The cluster environments created the S3 buckets without issue. Only the Vault configuration created additional folders under the bucket.

#### Which issue this PR fixes

fixes #8572

### CODE REVIEW

**Review by awisky — APPROVED** [review]

File: modules/cluster/storage.tf [hunk 1]
```diff
@@ -50,11 +59,20 @@ resource "aws_s3_bucket" "reports_jenkins_x" {
...
 
 resource "aws_s3_bucket_acl" "reports_jenkins_x" {
-  count  = var.enable_reports_storage ? 1 : 0
+  count  = var.enable_reports_storage && var.enable_acl ? 1 : 0
   bucket = aws_s3_bucket.reports_jenkins_x[0].bucket
   acl    = "private"
 }
 
+resource "aws_s3_bucket_ownership_controls" "reports_jenkins_x" {
+  count  = var.enable_reports_storage && !var.enable_acl ? 1 : 0
```
Comment (msvticket) [inline modules/cluster/storage.tf:68]:
```suggestion
  count  = var.enable_reports_storage && var.enable_acl ? 1 : 0
```

File: modules/cluster/storage.tf [hunk 2]
```diff
@@ -81,11 +99,20 @@ resource "aws_s3_bucket" "repository_jenkins_x" {
...
 
 resource "aws_s3_bucket_acl" "repository_jenkins_x" {
-  count  = var.enable_repository_storage ? 1 : 0
+  count  = var.enable_repository_storage && var.enable_acl ? 1 : 0
   bucket = aws_s3_bucket.repository_jenkins_x[0].bucket
   acl    = "private"
 }
 
+resource "aws_s3_bucket_ownership_controls" "repository_jenkins_x" {
+  count  = var.enable_repository_storage && !var.enable_acl ? 1 : 0
```
Comment (msvticket) [inline modules/cluster/storage.tf:108]:
```suggestion
  count  = var.enable_repository_storage && var.enable_acl ? 1 : 0
```

File: modules/vault/main.tf [hunk 3]
```diff
@@ -44,11 +44,20 @@ resource "aws_s3_bucket" "vault-unseal-bucket" {
...
 
 resource "aws_s3_bucket_acl" "vault-unseal-bucket" {
-  count  = local.create_vault_resources ? 1 : 0
+  count  = local.create_vault_resources && var.enable_acl ? 1 : 0
   bucket = aws_s3_bucket.vault-unseal-bucket[0].bucket
   acl    = "private"
 }
 
+resource "aws_s3_bucket_ownership_controls" "vault-unseal-bucket" {
+  count  = local.create_vault_resources && !var.enable_acl ? 1 : 0
```
Comment (msvticket) [inline modules/vault/main.tf:53]:
```suggestion
  count  = local.create_vault_resources && var.enable_acl ? 1 : 0
```

File: modules/cluster/storage.tf [hunk 4]
```diff
@@ -20,11 +20,20 @@ resource "aws_s3_bucket" "logs_jenkins_x" {
...
 
 resource "aws_s3_bucket_acl" "logs_jenkins_x" {
-  count  = var.enable_logs_storage ? 1 : 0
+  count  = var.enable_logs_storage && var.enable_acl ? 1 : 0
   bucket = aws_s3_bucket.logs_jenkins_x[0].bucket
   acl    = "private"
 }
 
+resource "aws_s3_bucket_ownership_controls" "logs_jenkins_x" {
+  count  = var.enable_logs_storage && !var.enable_acl ? 1 : 0
```
Comment (msvticket) [inline modules/cluster/storage.tf:29]:
```suggestion
  count  = var.enable_logs_storage && var.enable_acl ? 1 : 0
```

File: modules/backup/main.tf [hunk 5]
```diff
@@ -19,11 +19,20 @@ resource "aws_s3_bucket" "backup_bucket" {
...
 
 resource "aws_s3_bucket_acl" "backup_bucket" {
-  count  = var.enable_backup ? 1 : 0
+  count  = var.enable_backup && var.enable_acl ? 1 : 0
   bucket = aws_s3_bucket.backup_bucket[0].bucket
   acl    = "private"
 }
 
+resource "aws_s3_bucket_ownership_controls" "backup_bucket" {
+  count  = var.enable_backup && !var.enable_acl ? 1 : 0
```
Comment (msvticket) [inline modules/backup/main.tf:28]:
```suggestion
  count  = var.enable_backup && var.enable_acl ? 1 : 0
```

**Review by msvticket — CHANGES_REQUESTED** [review]
With the changes made by AWS the sufficient and appropriate change is to simply remove the acl.

From reading https://github.com/hashicorp/terraform-provider-aws/issues/28353 my understanding is that the aws_s3_bucket_ownership_controls resource is needed together with aws_s3_bucket_acl, not otherwise. So the count expression of the aws_s3_bucket_ownership_controls resources are wrong. This also means that the descriptions of the enable_acl variables need to be corrected.

> 
> With the new default security settings, a private ACL definition can likely be removed completely as the default security settings already ensure public access is restricted upon bucket creation.
> 
> However, if a private ACL is explicitly required, the following configuration splits bucket and ACL creation into separate API calls, with the requisite object ownership API call in between.

**Review by msvticket — APPROVED** [review]

**Pull request merged**